File: src/main/java/com/indiepost/service/SitemapServiceImpl.java
package com.indiepost.service;

import com.indiepost.dto.StaticPageDto;
import com.indiepost.dto.post.PostSummaryDto;
import com.indiepost.enums.Types.PostStatus;
import com.indiepost.model.Category;
import com.indiepost.repository.CategoryRepository;
import com.indiepost.repository.PostRepository;
import com.indiepost.repository.StaticPageRepository;
import cz.jiripinkas.jsitemapgenerator.WebPageBuilder;
import cz.jiripinkas.jsitemapgenerator.generator.SitemapGenerator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Created by jake on 17. 3. 21.
 */
@Service
@Transactional(readOnly = true)
public class SitemapServiceImpl implements SitemapService {
    private final PostRepository postRepository;
    private final CategoryRepository categoryRepository;
    private final StaticPageRepository staticPageRepository;

    @Autowired
    public SitemapServiceImpl(PostRepository postRepository, CategoryRepository categoryRepository, StaticPageRepository staticPageRepository) {
        this.postRepository = postRepository;
        this.categoryRepository = categoryRepository;
        this.staticPageRepository = staticPageRepository;
    }

    @Override
    public String buildSitemap() {
        SitemapGenerator sitemapGenerator = new SitemapGenerator("http://www.indiepost.co.kr");

        List<PostSummaryDto> posts = postRepository.findByStatus(
                PostStatus.PUBLISH,
                PageRequest.of(0, 9999, Sort.Direction.DESC, "publishedAt")
        );

        for (PostSummaryDto postSummaryDto : posts) {
            sitemapGenerator.addPage(new WebPageBuilder()
                    .name("post/" + postSummaryDto.getId())
                    .changeFreqDaily()
                    .priorityMax()
                    .build()
            );
        }
        for (Category category : categoryRepository.findAll()) {
            sitemapGenerator.addPage(new WebPageBuilder()
                    .name("category/" + category.getSlug())
                    .changeFreqDaily()
                    .priorityMax()
                    .build()
            );
        }
        List<StaticPageDto> pageList = staticPageRepository
                .find(PageRequest.of(0, 100, Sort.Direction.ASC, "displayOrder"))
                .getContent();
        for (StaticPageDto page : pageList) {
            sitemapGenerator.addPage(new WebPageBuilder()
                    .name("page/" + page.getSlug())
                    .changeFreqDaily()
                    .priorityMax()
                    .build()
            );
        }

        return sitemapGenerator.constructSitemapString();
    }
}

File: src/main/kotlin/com/indiepost/service/SitemapServiceImpl.kt
package com.indiepost.service

import com.indiepost.repository.PostRepository
import com.indiepost.repository.StaticPageRepository
import com.indiepost.repository.TagRepository
import cz.jiripinkas.jsitemapgenerator.WebPageBuilder
import cz.jiripinkas.jsitemapgenerator.generator.SitemapGenerator
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.inject.Inject

/**
 * Created by jake on 17. 3. 21.
 */
@Service
@Transactional(readOnly = true)
class SitemapServiceImpl @Inject constructor(
        private val postRepository: PostRepository,
        private val tagRepository: TagRepository,
        private val staticPageRepository: StaticPageRepository) : SitemapService {

    override fun buildSitemap(): String {
        val sitemapGenerator = SitemapGenerator("http://www.indiepost.co.kr")

        val posts = postRepository.findPublicPosts(
                PageRequest.of(0, 9999, Sort.Direction.DESC, "publishedAt")
        ).content

        for (postSummaryDto in posts) {
            sitemapGenerator.addPage(WebPageBuilder()
                    .name("post/" + postSummaryDto.id!!)
                    .changeFreqDaily()
                    .priorityMax()
                    .build()
            )
        }
        for (tag in tagRepository.findSelected()) {
            sitemapGenerator.addPage(WebPageBuilder()
                    .name("tag/$tag?primary=true")
                    .changeFreqDaily()
                    .priorityMax()
                    .build()
            )
        }
        val pageList = staticPageRepository
                .find(PageRequest.of(0, 100, Sort.Direction.ASC, "displayOrder"))
                .content
        for (page in pageList) {
            if (page.slug == null) {
                continue
            }
            sitemapGenerator.addPage(WebPageBuilder()
                    .name("page/${page.slug}")
                    .changeFreqDaily()
                    .priorityMax()
                    .build()
            )
        }

        return sitemapGenerator.constructSitemapString()
    }
}
